### ISSUE TITLE
Autocomplete on Query View (tables and fields)

### ISSUE THREAD

**Issue by danillos** [jakob/Postico#52]
Example:

![screen shot 2015-08-17 at 6 39 04 pm](https://cloud.githubusercontent.com/assets/54908/9317022/417ceb66-450f-11e5-84f2-19e550986de7.png)

**Comment by trapias** [comment]
Would be great!

**Comment by gbalytnikov** [comment]
+1

**Comment by kruser** [comment]
:+1:

**Comment by torbjornlunde** [comment]
+1

**Comment by radim** [comment]
+1

**Comment by davidesantangelo** [comment]
+1

**Comment by xarg** [comment]
:+1:

**Comment by kohidave** [comment]
:+1:

**Comment by thousandsofthem** [comment]
:+1:

**Comment by diddeb** [comment]
:heart:

**Comment by thathert** [comment]
:+1:

**Comment by winniwinter** [comment]
+1

**Comment by taehyunlim** [comment]
+1

**Comment by max-winderbaum** [comment]
+1

**Comment by dimfeld** [comment]
I would love to see autocomplete on SQL keywords as well. This doesn't come up too often but is nice to have in `psql` for being lazy with longer commands like `refresh materialized view`.

**Comment by akras14** [comment]
+1

**Comment by pbassut** [comment]
+1

**Comment by robinkeunen** [comment]
+1

**Comment by mex** [comment]
+1

**Comment by jpack482** [comment]
+1

**Comment by mcb** [comment]
:+1:

**Comment by sfotyler** [comment]
+1

**Comment by matejvelikonja** [comment]
+1

**Comment by Johnee** [comment]
+1

**Comment by stephenhowells** [comment]
👍🏼

**Comment by petervargaofficial** [comment]
+1 pls, pls add this!

**Comment by christophe-f** [comment]
+1

**Comment by zak01011996** [comment]
+1

**Comment by wojtek-fliposports** [comment]
+1

**Comment by rajofchennai** [comment]
+1

**Comment by dzirtbry** [comment]
+1

**Comment by mattpelaggi** [comment]
+1

**Comment by mattpelaggi** [comment]
+1

**Comment by gportillo** [comment]
+1

**Comment by RickyMcMaster** [comment]
+1

**Comment by brego** [comment]
+1

**Comment by kdelrosso** [comment]
+1

**Comment by luisfontes** [comment]
+1

**Comment by gerbsen** [comment]
+1

**Comment by guillemettem** [comment]
+1

**Comment by chano** [comment]
+1

**Comment by joshmanders** [comment]
Hey, Github has a reactions feature, instead of spamming people's emails with +1 how about just reacting to the original post to show you're agreeing.

**Comment by pbassut** [comment]
@joshmanders You can always unsubscribe from an issue. 

The problem with using reactions is that it doesn't send an email thus people(or the owner) won't know that feature is wanted. Meaning you should check the issue every time to see it's demand.

In the case of a private project like this, it's kind of okay to use reactions. But for open-source projects, I don't think so.

**Comment by jakob** [comment]
@pbassut you are right in general, but in this case I do look at reactions. I have a script that fetches issues from Github and ranks them by reactions + comments. (This issue is by far the highest ranked). Maybe I should make that public somewhere...

So yeah, maybe it would be better to use reactions -- i don't want people to unsubscribe from issues, because then they won't be notified when I post news about an issue.

**Comment by jurati** [comment]
+1

**Comment by joshmanders** [comment]
@pbassut the whole point of subscribing to issues is so I can be updated on pertinent information regarding that issue, not about an +1 or an emoji.

**Comment by maxibor** [comment]
+1

**Comment by mitchelllisle** [comment]
+1

**Comment by benoittgt** [comment]
+1 and BTW thanks for Postico. Love using it.

**Comment by ineffyble** [comment]
+1

**Comment by schmitzdenis** [comment]
+1

**Comment by hmaciasc** [comment]
+1

**Comment by nomut** [comment]
+1 Dream will come true

**Comment by tallavi** [comment]
+2

**Comment by digiconz** [comment]
+1

**Comment by ccbrandenburg** [comment]
+1

**Comment by avnersentinelone** [comment]
+1

**Comment by valeria-a** [comment]
+1

**Comment by zankbennett** [comment]
+1

**Comment by benoittgt** [comment]
That's nice to see all of you interested in this feature but please as mention in Readme

> Please use the "thumbs up" button to show support for a feature request. Avoid posting "+1" comments, since those will be forwarded to everyone watching the repo.

Cheers !

**Comment by jakob** [comment]
I just tried locking the issue to block the +1 comments, but then reactions are disabled too. Sorry for the noise.

**Comment by gerhat** [comment]
+1

**Comment by donbeave** [comment]
+1

**Comment by andrebrov** [comment]
+1

**Comment by alexpetrov84** [comment]
+1
would be great to have this!

**Comment by sphinks** [comment]
+1

**Comment by dubininss** [comment]
+1

**Comment by CanCeylan** [comment]
+1

**Comment by Spencerkt** [comment]
+1

**Comment by reganjohnson** [comment]
+100

**Comment by jakob** [comment]
Big news! We finally have something to show for the most requested feature in Postico!

Over the last few weeks, @martinjankoehler has implemented a basic autocomplete function in Postico.

Please download build 1838 from https://eggerapps.at/postico/builds/

We'd love to hear your feedback!

PS: I'm now going to delete all those +1 comments to make the discussion easier to follow. Please use reactions to show support. Use the "Subscribe" button in the sidebar if you'd like to receive notifications for this issue.

**Comment by jakob** [comment]
It looks like Github will unsubscribe users when I delete their "+1" comments. Please re-subscribe to notifications for this issue on Github if you'd like to follow the discussion.

**Comment by gerbsen** [comment]
I just downloaded 1838 but I cannot see any autocompletion? What am I doing wrong?

**Comment by jakob** [comment]
@gerbsen Did you try pressing `ESC`? Autocomplete doesn't trigger automatically (yet).

**Comment by gerbsen** [comment]
No, how would I know that? :)  Do you have documentation where and when it works?

I just tried `select * from proj + ESC` suggests right project table
I just tried `select * from project WHERE + ESC` nothing showing
I just tried `select * from project WHERE project. + ESC` nothing showing

Cheers, and thanks for your work and @martinjankoehler

**Comment by jakob** [comment]
There are no docs yet. If something isn't obvious we need to fix it.

Thanks for your feedback. Just curious: Did you expect autocomplete to trigger automatically, or did you try other keys beside ESC?

**Comment by kruser** [comment]
I think `ctrl+space` is the most natural autocomplete keystroke.

I have the same issues at @gerbsen where this only auto completes my table name, not the column names. Promising stuff though!

**Comment by iovis9** [comment]
I don't know if it's because I'm a vim user, but I've never used `Esc` for autocompletion. I would have tried `Tab` and `ctrl+space`.

**Comment by gerbsen** [comment]
for mac it would be cmd + space and if I type "." for an object/table-(reference)

**Comment by danillos** [comment]
About the shortcut, I suggest to have a key binding on the preferences for it, every person has your own preferences.

Would be good to trigger it when hit `.` after a table name too.

**Comment by jzk** [comment]
I was expecting similar behavior like https://github.com/dbcli/pgcli

**Comment by meeDamian** [comment]
`⌘ + space` on Macs is usually reserved for Spotlight

IntelliJ and many other IDEs use `^ + space` for suggestions, which I think, should work best.

Ideally, suggestions would just open on typing, but for it to be more useful than annoying, understanding the query being typed would probably be necessary.

**Comment by jakob** [comment]
Thanks for all the comments! 

I wasn't aware that `^ space` is such a common shortcut for autocomplete -- even Xcode uses it (in addition to ESC). I've now added a menu item titled "Show Completions" with the shortcut ctrl-space. If you download [build 1848](https://eggerapps.at/postico/builds/) you can now use the following keys to trigger autocomplete:

- ctrl-space
- esc
- F5

If you want to customize the keyboard shortcut, you can go to System Preferences -> Keyboard -> Shortcuts -> App Shortcuts and add a custom shortcut for the command "Show Completions"

Right now Autocomplete requires at least one character to show completions; Martin is working on removing this requirement (so that eg. column names are shown even if you haven't typed anything yet). We also want to add some context sensitivity, so that Postico suggest the correct columns after typing `tablename.`

As for triggering autocomplete automatically, we had the same thoughts as @meeDamian: Without understanding the query, there's little point in showing a seemingly infinite list of completions. Our goal is absolutely to have Postico understand the query, and offer appropriate completions automatically -- but it's going to take us some time to get there.

**Comment by zak01011996** [comment]
Nice, finally it make sense :-) 

Waiting for the final release and after that I will buy a license.

Iskander Khabirov
zak01011996@gmail.com

**Comment by kruser** [comment]
Looking better @jakob. FYI, it doesn't seem to work for Redshift.

**Comment by jakob** [comment]
Hey everyone!

@martinjankoehler has been hard at work, and Autocomplete works a lot better now!

- Autocomplete now works on Redshift (thanks for the reminder @kruser)
- Autocomplete automatically triggers when you type a period after a table name (and shows relevant results)
- You can now trigger autocomplete even when you haven't typed part of the name yet
- ... and lots and lots of minor improvements

Please give it a try. I think it's already pretty usable.

Please download build 1915 from https://eggerapps.at/postico/builds/

**Comment by thousandsofthem** [comment]
looks good! i like it. no issues so far

sidenotes:
* `tab` hotkey would be nice (customizable hotkeys maybe?)
* autocomplete can be made more smart in the future, e.g. after 
```
WHERE attrname <autocomplete>
WHERE attrname IS <autocomplete>
```
do not suggest attributes again but, instead, typical constructions like `IS NULL`/`IS NOT NULL` etc

* for built-in functions popup could show short description from the docs (db/version specific?)

**Comment by xuefeng-huang** [comment]
Hi, 
I am using trial Version 1.2.3 on Mac. Could anyone give me a hint how I can enable autocomplete function? Thanks

**Comment by jakob** [comment]
@xuefeng-huang Autocomplete is still under development. Please download the latest build of the autocomplete branch from https://eggerapps.at/postico/builds/

**Comment by zak01011996** [comment]
Really waiting for stable release to buy a license!

Iskander Khabirov
zak01011996@gmail.com

**Comment by martinjankoehler** [comment]
Hey everyone,

changes in [Build 1978](https://eggerapps-downloads.s3-eu-west-1.amazonaws.com/postico-1978.zip) include:
-  automatically add parens when autocompleting SQL functions
- hitting tab in the context of a token prefix triggers autocompletion
- autocompletion error panel now non-modal
- suggestion ranking improvements
- minor cosmetic improvements

Best regards, Martin

**Comment by jakob** [comment]
Hi!

We've released Postico 1.3 a couple of days ago, which adds autocomplete.

I'd like to thank everyone in this issue for suggestions, feedback, and especially for testing prerelease builds. It's awesome to get so much response.

I'm closing this issue now. If you have any suggestions for improving Autocomplete, please open a new issue!

Thanks!

**Comment by auxo86** [comment]
Hi~I am a new user of pgadmin under windows.
It's awesome.
However, Ctrl + Space is the default keyboard shortcut for switch IM method.
Autocomplete feature does not work in my win10 desktop.
I have searched two days for change the setting of Autocomplete Keyboard shortcut.
Any help will be appreciated.
Many thanks.
